### ISSUE TITLE
Array `map`/`filter`/`reduce` fail with typed-array variables and functions

### ISSUE THREAD

**Issue by rblopes** [godotengine/godot#72566]
### Godot version

v4.0.beta17.official [c40020513]

### System information

KDE neon 22.04 (Ubuntu-derived)

### Issue description

After beta.17, trying to assign the result of `Array#map`, `Array#filter` or `Array#reduce` to typed-array variables, or returning their values as the result of typed-array functions fail.

The following sample code demonstrates the issue:

```gdscript
@tool
extends EditorScript

func test1() -> Array[int]:
	return range(5).map(func(n): return n) # → [0, 1, 2, 3, 4]

func test2() -> Array[int]:
	var result: Array[int] = range(10).filter(func(n): return n % 2 == 0) # → [0, 2, 4, 6, 8]
	return result

func _run() -> void:
	print(test1())
	print(test2())
```

![Screenshot_20230202_013030](https://user-images.githubusercontent.com/61425/216232027-2cdeb150-c964-4d02-afd5-1f2e1658ac86.png)

Casting the returned values doesn't work either.

### Steps to reproduce

Assign the result of `Array#map`, `Array#filter` or `Array#reduce` to a variable whose type is a qualified typed-array.

### Minimal reproduction project

N/A

**Comment by dalexeev** [comment]
> ```gdscript
> func test2() -> Array[int]:
> 	var result: Array[int] = range(10).filter(func(n): return n % 2 == 0) # → [0, 2, 4, 6, 8]
> 	return result
> ```

The problem is that `Array` methods return a generic `Array` (`Array[Variant]`). For example `Array[T].filter()` returns an `Array`, not an `Array[T]`.

For now, you can use `Array.assign()` method:

```gdscript
func test() -> Array[int]:
    var result: Array[int]
    result.assign(range(10).filter(func(n): return n % 2 == 0))
    return result
```

In the future, if #71336 is accepted, you will be able to do:


```gdscript
func test() -> Array[int]:
    return Array[int](range(10).filter(func(n): return n % 2 == 0))
```

And, just in case, you can `range(0, 10, 2)` instead of `range(10).filter(func(n): return n % 2 == 0)`. But I think you know this, and this is just an example. :)

**Comment by rblopes** [comment]
Ok, looking at it again, I've found this code snippet works as intended, no warnings or errors issued.

```gdscript
@tool
extends EditorScript

var some_var: Array[int] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

func test() -> Array[int]:
	return some_var.filter(func(n): return n >= 5)

func _run() -> void:
	print(test()) # → [5, 6, 7, 8, 9]
```

My guess is that the GDScript parser infers the filtered array type from `some_var`.

By the way, replacing the `some_var` init expression with `range()` crashes the editor when I run that script.

**Comment by kleonc** [comment]
`Array.filter` preserves the typing, it returns an array typed the same as the original array. Hence:

```gdscript
var some_var: Array[int] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
var result: Array[int] = some_var.filter(func(n): return n >= 5)
```

works fine. The problem with:
```gdscript
var result: Array[int] = range(10).filter(func(n): return n % 2 == 0)
```
is that `range` currently returns an untyped `Array` and hence the types mismatch. #67025 should fix this.

---

Regarding `Array.map`: it returns an untyped `Array`, it doesn't preserve the typing of the original array. That's intended, as elements can be mapped to any type.

To potentially make `map` return a typed array based on the return type of the passed `Callable`, the `Callable` would need to provide a way to obtain its return type in the first place (currently it's not supported).

---

> (By the way, replacing the `some_var` init expression with `range()` crashes the editor when I run that script.)

Example with no `range` involved:

```gdscript
@tool
extends EditorScript

var a: Array = []
var b: Array[int] = a # (1) res://path.gd:5 - Trying to assign an array of type "Array" to a variable of type "Array[int]".

func _run() -> void: # (0) Execute in editor.
	print(b) # (2) <null>
	print(b.size()) # (3) Crash.
```

@vnen @vonagam Is this expected? :thinking:

**Comment by vonagam** [comment]
```gdscript
var a: Array = []
var b: Array[int] = a # error
```
Yes, `a` points to an untyped array. Cannot assign an untyped array to a typed array variable.

**Comment by kleonc** [comment]
@vonagam Thanks. And sorry, I should have been more precise as I wasn't asking specifically about the reason of the error, I understand that part. I was rather asking about the nature of the error / post-error behavior. Like e.g. if it should be a fatal error disallowing further execution (and thus potential crash because of the `<null>` value), or maybe if it should just error and assign an empty array instead?

For comparison when conversion of the elements fails it also errors but results in an empty array instead:
```gdscript
@tool
extends EditorScript

var integer = "0"
var array_of_integers = [integer]

var a: Array[int] = [integer]
var b: Array[int] = array_of_integers

func _run():
	print(a)
	print(b)
```
Output (v4.0.beta17.official [c40020513]):
```
  Unable to convert array index 0 from 'String' to 'int'.
  res://ES.gd:8 - Trying to assign an array of type "Array" to a variable of type "Array[int]".
[]
<null>
```

Again, I kinda understand why it currently gives such output. Just not sure whether it's truly intended. Shouldn't both cases result in unitialized value / empty array? :thinking:

**Comment by dalexeev** [comment]
> `Array.filter` preserves the typing, it returns an array typed the same as the original array.

Yes, you're right, that's a good clarification. **But the static type is not inferred**, it's a generic `Array`, despite the fact that the array (object in memory) is typed at runtime.

**Comment by vonagam** [comment]
> Shouldn't both cases result in unitialized value / empty array?

In my opinion both cases should result in a fatal error. It is the same as assigning `String` value to `int` variable, everything that will happen after such wrong deed is undefined. For example, if it was a function that deletes files and an array of strings was supposed to be whitelist, to prevent files from being deleted, one wouldn't want it to proceed with an empty array.

**Comment by KeatsPeeks** [comment]
Fyi, as a new user, I lost time trying to understand this while trying to use .map() on a typed array. Honestly it feels like Java 1.4 all over again when generics where introduced... Which is 20 years-old now.

**Comment by MikeSchulze** [comment]
Hi I ran into the same problem today, is there an update here?
It's really inconvenient to use an extra assignment instead of directly returning the result of a map() function.
The array map function is not usable for typed arrays. This is a bug and should definitely be fixed. The map function should return the type of the mapping function used.
Yes, the workaround works, but makes the code more complex than necessary, this negates the benefits of the map function again. 

```gdscritp
var values :Array[String] = ["_1", "_2", "_3"]

# fails at runtime
func convert_to_id1() -> Array[int]:
	return values.map(_to_id)

# works with the assign workaround
func convert_to_id2() -> Array[int]:
	var result :Array[int] = []
	result.assign(values.map(_to_id))
	return result


func _to_id(value :String) -> int:
	return value.replace("_", "").to_int()
```

**Comment by portwatcher** [comment]
The only workaround for me is to create a new typed array manually:
```gdscript
	var recipes = craft_recipes.filter(
		func(recipe: RecipeItem):
                        # do something
			return true
	).map(
		func(recipe: RecipeItem) -> RecipeItem:
			# do something
			return recipe
	)
	
	var ret_recipes: Array[RecipeItem]
	
	# a godot bug:
	# https://github.com/godotengine/godot/issues/72566
	# https://github.com/godotengine/godot/pull/71336
	for recipe in recipes:
		ret_recipes.push_back(recipe)
	
	return ret_recipes
```

**Comment by limbonaut** [comment]
@portwatcher 
You can use Array.assign(), which should be faster.

**Comment by portwatcher** [comment]
Yeah but I only have like 20 recipes and `Array.assign` is not visibly
straightforward to me

Serhii Snitsaruk ***@***.***> 于2023年12月30日周六 01:42写道：

**Comment by portwatcher** [comment]
> @portwatcher You can use Array.assign(), which should be faster.

Yeah but I only have like 20 recipes and `Array.assign` is not visibly straightforward to me

**Comment by AThousandShips** [comment]
How do you mean? It's just:
```gdscript
	var recipes = craft_recipes.filter(
		func(recipe: RecipeItem):
                        # do something
			return true
	).map(
		func(recipe: RecipeItem) -> RecipeItem:
			# do something
			return recipe
	)
	
	var ret_recipes: Array[RecipeItem]
	
	ret_recipes.assign(recipes)
	
	return ret_recipes
```

**Comment by Oblepikha** [comment]
If you desperately need a one-liner there is a workaround

```gdscript
func do(array: Array[ClassA]) -> Array[ClassB]:
	#var result: Array[ClassB] = []
	#result.assign(array.map(func(element: ClassA) -> ClassB: return ClassB.new(element)))
	#return result
	return Array(array.map(func(element: ClassA) -> ClassB: return ClassB.new(element)), TYPE_OBJECT, "RefCounted", ClassB)
```

**Comment by badunius** [comment]
``

> How do you mean? It's just:
> 
> ```gdscript
> 	var ret_recipes: Array[RecipeItem]
> 	
> 	ret_recipes.assign(recipes)
> 	
> 	return ret_recipes
> ```

why this even here?
when `map` is called with a typed `Callable` it supposed to return an `Array[typeof ReturnType<Callable>]`
actually it should work like this all the time, since `ReturnType` of untyped `Callable` is `Variant`
then we're going to have an experience consistent with what we have now and actually respect typing

for now it's easier to just drop typing in such cases

**Comment by codevogel** [comment]
```
func to_path() -> Array[Vector2]:
    return _tiles.map(func(tile) -> Vector2: return tile.world_position)

**Comment by mikemike37** [comment]
I tripped up here as well. I was expecting cases like this to work:
`var array_of_ints: Array[int] = ["1","2","3"].map(func (i: String) -> int: return int(i))`

Thanks to the replies here, I was able to find a solution using assign. In case it helps anyone else, I made this utility function to map in a type-safe way:
```
func map_type_safe(array_to_map: Array[Variant], result_typed_byref, callable) -> void:
	var array_untyped = array_to_map.map(callable)
	result_typed_byref.assign(array_untyped)

func example_usage() -> Array[int]:
	var array_to_map: Array[Variant] = ["1","2","3"]
	var result_typed_byref: Array[int]
	var callable = func (i: String) -> int: return int(i)
	map_type_safe(array_to_map, result_typed_byref, callable)
	return result_typed_byref
```

... though I would still like if map returned a typed array when the callable is typed to avoid this shenanigans!

**Comment by mikemike37** [comment]
I tripped up here as well. For anyone else that struggled a bit to grasp the workaround for `map`, replace this:

`var array_of_ints: Array[int] = ["1","2","3"].map(func (i: String) -> int: return int(i))`

with this:
```
var array_of_ints: Array[int]
array_of_ints.assign(["1","2","3"].map(func (i: String) -> int: return int(i)))
```
+1 from me on suggestion that typed callables used in `map` should return typed arrays.

**Comment by zimonitrome** [comment]
@mikemike37 that's a nice workaround but does not appear to work on class properties :/
